StaTheory is not supported in Visual Studio Runner

Example

```cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Collections.Generic;
using WinForms.Common.Tests;
using Xunit;

namespace System.Windows.Forms.Tests
{
    public class WebBrowserTests
    {
        public static IEnumerable<object[]> Parent_Set_TestData()
        {
           // yield return new object[] { null };
            yield return new object[] { new Control() };
        }

        [StaTheory]
        [MemberData(nameof(Parent_Set_TestData))]
        public void WebBrowser_Parent_Set_GetReturnsExpected(Control parent)
        {
            var browser = new WebBrowser
            {
                Parent = parent
            };
            Assert.Same(parent, browser.Parent);
            Assert.Null(browser.ActiveXInstance);
            Assert.False(browser.IsHandleCreated);
        }
    }
}
```

```
[9/23/2019 7:22:30.780 PM Informational] ---------- Discovery started ----------
[9/23/2019 7:22:33.339 PM Informational] [xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.1 (32-bit Desktop .NET 4.0.30319.42000)
[9/23/2019 7:22:34.552 PM Informational] [xUnit.net 00:00:01.26]   Discovering: WinformsTest
[9/23/2019 7:22:35.854 PM Informational] [xUnit.net 00:00:02.40]   Discovered:  WinformsTest
[9/23/2019 7:22:35.855 PM Informational] [xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.1 (32-bit Universal Windows)
[9/23/2019 7:22:36.060 PM Informational] ========== Discovery finished: 85 tests found (0:00:05.2286991) ==========
[9/23/2019 7:22:36.116 PM Informational] ---------- Run started ----------
[9/23/2019 7:22:38.520 PM Informational] [xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.1 (32-bit Desktop .NET 4.0.30319.42000)
[9/23/2019 7:22:40.388 PM Error] [xUnit.net 00:00:01.88] WinformsTest: Test case System.Windows.Forms.Tests.WebBrowserTests.WebBrowser_Parent_SetWithHandle_GetReturnsExpected failed to deserialize: Test case deserialization failure: System.InvalidOperationException: Could not de-serialize type 'Xunit.Sdk.UITheoryTestCase' because it lacks a parameterless constructor.
   at Xunit.Serialization.XunitSerializationInfo.DeserializeSerializable(Type type, String serializedValue) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 213
   at Xunit.Serialization.XunitSerializationInfo.Deserialize(Type type, String serializedValue) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 110
   at Xunit.Sdk.SerializationHelper.Deserialize[T](String serializedValue) in C:\Dev\xunit\xunit\src\common\SerializationHelper.cs:line 40
   at Xunit.Sdk.XunitTestFrameworkExecutor.Deserialize(String value) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\XunitTestFrameworkExecutor.cs:line 84
   at Xunit.Sdk.TestCaseBulkDeserializer.Deserialize(ITestFrameworkDiscoverer discoverer, ITestFrameworkExecutor executor, String serialization) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\TestCaseBulkDeserializer.cs:line 33
[9/23/2019 7:22:40.390 PM Error] [xUnit.net 00:00:01.88] WinformsTest: Test case System.Windows.Forms.Tests.WebBrowserTests.WebBrowser_Parent_Set_GetReturnsExpected failed to deserialize: Test case deserialization failure: System.InvalidOperationException: Could not de-serialize type 'Xunit.Sdk.UITheoryTestCase' because it lacks a parameterless constructor.
   at Xunit.Serialization.XunitSerializationInfo.DeserializeSerializable(Type type, String serializedValue) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 213
   at Xunit.Serialization.XunitSerializationInfo.Deserialize(Type type, String serializedValue) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 110
   at Xunit.Sdk.SerializationHelper.Deserialize[T](String serializedValue) in C:\Dev\xunit\xunit\src\common\SerializationHelper.cs:line 40
   at Xunit.Sdk.XunitTestFrameworkExecutor.Deserialize(String value) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\XunitTestFrameworkExecutor.cs:line 84
   at Xunit.Sdk.TestCaseBulkDeserializer.Deserialize(ITestFrameworkDiscoverer discoverer, ITestFrameworkExecutor executor, String serialization) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\TestCaseBulkDeserializer.cs:line 33
[9/23/2019 7:22:40.416 PM Informational] [xUnit.net 00:00:01.89]   Starting:    WinformsTest
[9/23/2019 7:22:47.282 PM Informational] [xUnit.net 00:00:08.77]   Finished:    WinformsTest
[9/23/2019 7:22:47.807 PM Informational] ========== Run finished: 628 tests run (0:00:11.4979809) ==========
```

@RussKie I tried your repro, but couldn't get the solution to build. I get these errors:

```
1>------ Skipped Build: Project: Accessibility-version ------
1> 
2>------ Skipped Build: Project: InternalUtilitiesForTests ------
2> 
3>------ Skipped Build: Project: Microsoft.Dotnet.Winforms.ProjectTemplates ------
3> 
4>------ Skipped Build: Project: Accessibility (src\Accessibility\Accessibility) ------
4> 
5>------ Skipped Build: Project: System.Windows.Forms ------
5> 
6>------ Skipped Build: Project: System.Windows.Forms.Design.Editors ------
6> 
7>------ Build started: Project: System.Windows.Forms.IntegrationTests.Common, Configuration: Debug Any CPU ------
8>------ Skipped Build: Project: System.Windows.Forms.Tests ------
8> 
9>------ Build started: Project: AccessibilityTests, Configuration: Debug Any CPU ------
10>------ Skipped Build: Project: System.Windows.Forms.Design ------
10> 
11>------ Skipped Build: Project: System.Drawing.Design.Facade ------
11> 
12>------ Skipped Build: Project: System.Windows.Forms.Design.Editors.Tests ------
12> 
13>------ Skipped Build: Project: System.Design.Facade ------
13> 
14>------ Skipped Build: Project: System.Drawing.Facade ------
14> 
15>------ Skipped Build: Project: System.Windows.Forms.Design.Tests ------
15> 
16>------ Skipped Build: Project: Microsoft.Private.Winforms ------
16> 
7>CSC : error CS0006: Metadata file 'D:\git\winforms\artifacts\bin\Accessibility\Debug\netcoreapp5.0\Accessibility.dll' could not be found
7>CSC : error CS0006: Metadata file 'D:\git\winforms\artifacts\bin\System.Windows.Forms\Debug\netcoreapp5.0\ref\System.Windows.Forms.dll' could not be found
7>Done building project "System.Windows.Forms.IntegrationTests.Common.csproj" -- FAILED.
17>------ Skipped Build: Project: WinformsControlsTest ------
17> 
18>------ Skipped Build: Project: MauiButtonTests ------
18> 
19>------ Build started: Project: MauiComboBoxTests, Configuration: Debug Any CPU ------
20>------ Skipped Build: Project: System.Windows.Forms.IntegrationTests ------
20> 
19>CSC : error CS0006: Metadata file 'D:\git\winforms\artifacts\bin\Accessibility\Debug\netcoreapp5.0\Accessibility.dll' could not be found
19>CSC : error CS0006: Metadata file 'D:\git\winforms\artifacts\bin\System.Windows.Forms\Debug\netcoreapp5.0\ref\System.Windows.Forms.dll' could not be found
19>CSC : error CS0006: Metadata file 'D:\git\winforms\artifacts\bin\System.Windows.Forms.IntegrationTests.Common\Debug\netcoreapp5.0\System.Windows.Forms.IntegrationTests.Common.dll' could not be found
19>Done building project "MauiComboBoxTests.csproj" -- FAILED.
21>------ Skipped Build: Project: System.Windows.Forms.Maui.IntegrationTests ------
21> 
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018: The "GenerateResource" task failed unexpectedly.
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018: System.NullReferenceException: Object reference not set to an instance of an object.
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.Tasks.ProcessResourceFiles.get_HaveSystemResourcesExtensionsReference()
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.Tasks.ProcessResourceFiles.WriteBinaryResources(ReaderInfo reader, String filename)
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.Tasks.ProcessResourceFiles.WriteResources(ReaderInfo reader, String filename)
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.Tasks.ProcessResourceFiles.ProcessFile(String inFile, String outFileOrDir)
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.Tasks.ProcessResourceFiles.Run(TaskLoggingHelper log, ITaskItem[] assemblyFilesList, List`1 inputs, List`1 satelliteInputs, List`1 outputs, Boolean sourcePath, Boolean usePreserializedResources, String language, String namespacename, String resourcesNamespace, String filename, String classname, Boolean publicClass, Boolean extractingResWFiles, String resWOutputDirectory)
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.Tasks.GenerateResource.Execute()
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
9>Done building project "AccessibilityTests.csproj" -- FAILED.
========== Build: 0 succeeded, 3 failed, 0 up-to-date, 18 skipped ==========
```

Are you doing it from a cli or from VS?
Please try starting it from build-local.ps1

On Sat, Oct 5, 2019, 1:08 AM Andrew Arnott <notifications@github.com> wrote:

> @RussKie <https://github.com/RussKie> I tried your repro, but couldn't
> get the solution to build. I get these errors:
>
> 1>------ Skipped Build: Project: Accessibility-version ------
> 1>
> 2>------ Skipped Build: Project: InternalUtilitiesForTests ------
> 2>
> 3>------ Skipped Build: Project: Microsoft.Dotnet.Winforms.ProjectTemplates ------
> 3>
> 4>------ Skipped Build: Project: Accessibility (src\Accessibility\Accessibility) ------
> 4>
> 5>------ Skipped Build: Project: System.Windows.Forms ------
> 5>
> 6>------ Skipped Build: Project: System.Windows.Forms.Design.Editors ------
> 6>
> 7>------ Build started: Project: System.Windows.Forms.IntegrationTests.Common, Configuration: Debug Any CPU ------
> 8>------ Skipped Build: Project: System.Windows.Forms.Tests ------
> 8>
> 9>------ Build started: Project: AccessibilityTests, Configuration: Debug Any CPU ------
> 10>------ Skipped Build: Project: System.Windows.Forms.Design ------
> 10>
> 11>------ Skipped Build: Project: System.Drawing.Design.Facade ------
> 11>
> 12>------ Skipped Build: Project: System.Windows.Forms.Design.Editors.Tests ------
> 12>
> 13>------ Skipped Build: Project: System.Design.Facade ------
> 13>
> 14>------ Skipped Build: Project: System.Drawing.Facade ------
> 14>
> 15>------ Skipped Build: Project: System.Windows.Forms.Design.Tests ------
> 15>
> 16>------ Skipped Build: Project: Microsoft.Private.Winforms ------
> 16>
> 7>CSC : error CS0006: Metadata file 'D:\git\winforms\artifacts\bin\Accessibility\Debug\netcoreapp5.0\Accessibility.dll' could not be found
> 7>CSC : error CS0006: Metadata file 'D:\git\winforms\artifacts\bin\System.Windows.Forms\Debug\netcoreapp5.0\ref\System.Windows.Forms.dll' could not be found
> 7>Done building project "System.Windows.Forms.IntegrationTests.Common.csproj" -- FAILED.
> 17>------ Skipped Build: Project: WinformsControlsTest ------
> 17>
> 18>------ Skipped Build: Project: MauiButtonTests ------
> 18>
> 19>------ Build started: Project: MauiComboBoxTests, Configuration: Debug Any CPU ------
> 20>------ Skipped Build: Project: System.Windows.Forms.IntegrationTests ------
> 20>
> 19>CSC : error CS0006: Metadata file 'D:\git\winforms\artifacts\bin\Accessibility\Debug\netcoreapp5.0\Accessibility.dll' could not be found
> 19>CSC : error CS0006: Metadata file 'D:\git\winforms\artifacts\bin\System.Windows.Forms\Debug\netcoreapp5.0\ref\System.Windows.Forms.dll' could not be found
> 19>CSC : error CS0006: Metadata file 'D:\git\winforms\artifacts\bin\System.Windows.Forms.IntegrationTests.Common\Debug\netcoreapp5.0\System.Windows.Forms.IntegrationTests.Common.dll' could not be found
> 19>Done building project "MauiComboBoxTests.csproj" -- FAILED.
> 21>------ Skipped Build: Project: System.Windows.Forms.Maui.IntegrationTests ------
> 21>
> 9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018: The "GenerateResource" task failed unexpectedly.
> 9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018: System.NullReferenceException: Object reference not set to an instance of an object.
> 9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.Tasks.ProcessResourceFiles.get_HaveSystemResourcesExtensionsReference()
> 9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.Tasks.ProcessResourceFiles.WriteBinaryResources(ReaderInfo reader, String filename)
> 9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.Tasks.ProcessResourceFiles.WriteResources(ReaderInfo reader, String filename)
> 9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.Tasks.ProcessResourceFiles.ProcessFile(String inFile, String outFileOrDir)
> 9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.Tasks.ProcessResourceFiles.Run(TaskLoggingHelper log, ITaskItem[] assemblyFilesList, List`1 inputs, List`1 satelliteInputs, List`1 outputs, Boolean sourcePath, Boolean usePreserializedResources, String language, String namespacename, String resourcesNamespace, String filename, String classname, Boolean publicClass, Boolean extractingResWFiles, String resWOutputDirectory)
> 9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.Tasks.GenerateResource.Execute()
> 9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
> 9>C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(3056,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
> 9>Done building project "AccessibilityTests.csproj" -- FAILED.
> ========== Build: 0 succeeded, 3 failed, 0 up-to-date, 18 skipped ==========
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AArnott/Xunit.StaFact/issues/30?email_source=notifications&email_token=ABBTEXVKA2FDG3XROPKE4IDQM5L7BA5CNFSM4IZP65VKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAL6QAY#issuecomment-538437635>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBTEXWBJCTXJK7DDDJIBCDQM5L7BANCNFSM4IZP65VA>
> .
>

--- a/src/Xunit.StaFact/Sdk/UITheoryTestCase.cs
+++ b/src/Xunit.StaFact/Sdk/UITheoryTestCase.cs
@@ -3,21 +3,43 @@
 
 namespace Xunit.Sdk
 {
+    using System;
+    using System.ComponentModel;
     using System.Threading;
     using System.Threading.Tasks;
     using Xunit.Abstractions;
 
     public class UITheoryTestCase : XunitTheoryTestCase
     {
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [Obsolete("Called by the de-serializer; should only be called by deriving classes for de-serialization purposes")]
+        public UITheoryTestCase()
+        {
+        }
+
         public UITheoryTestCase(UITestCase.SyncContextType synchronizationContextType, IMessageSink diagnosticMessageSink, TestMethodDisplay defaultMethodDisplay, TestMethodDisplayOptions defaultMethodDisplayOptions, ITestMethod testMethod)
             : base(diagnosticMessageSink, defaultMethodDisplay, defaultMethodDisplayOptions, testMethod)
         {
             this.SynchronizationContextType = synchronizationContextType;
         }
 
-        internal UITestCase.SyncContextType SynchronizationContextType { get; }
+        internal UITestCase.SyncContextType SynchronizationContextType { get; private set; }
 
         public override Task<RunSummary> RunAsync(IMessageSink diagnosticMessageSink, IMessageBus messageBus, object[] constructorArguments, ExceptionAggregator aggregator, CancellationTokenSource cancellationTokenSource)
             => new UITheoryTestCaseRunner(this, this.DisplayName, this.SkipReason, constructorArguments, diagnosticMessageSink, messageBus, aggregator, cancellationTokenSource).RunAsync();
+
+        public override void Deserialize(IXunitSerializationInfo data)
+        {
+            base.Deserialize(data);
+
+            this.SynchronizationContextType = (UITestCase.SyncContextType)Enum.Parse(typeof(UITestCase.SyncContextType), data.GetValue<string>("SyncContextType"));
+        }
+
+        public override void Serialize(IXunitSerializationInfo data)
+        {
+            base.Serialize(data);
+
+            data.AddValue("SyncContextType", this.SynchronizationContextType.ToString());
+        }
     }
 }

